### PR TITLE
fixes bug 1352385 - Use active versions if 0 featured versions

### PR DESCRIPTION
**Pull request by peterbe**
This basically makes the home page, Top Crashers, Crashes Per Day work again if the product *has* active versions but *no featured* versions.

### CODE REVIEW

**Comment by peterbe** [comment]
screen shot 2017-03-31 at 8 58 11 am
It works again. Yay!

File: webapp-django/crashstats/base/jinja2/crashstats_base.html [hunk 1]
```diff
@@ -59,26 +59,30 @@
...
+                    {% for pv in pvs %}
                         <option
                             value="{{ pv.version }}"
                             {% if pv.version == version %}selected{% endif %}
                         >{{ pv.version }}</option>
-                        {% endif %}
-                    {% endfor  %}
+                    {% endfor %}
                     </optgroup>
+                    {% endif %}
```
Comment (peterbe) [inline webapp-django/crashstats/base/jinja2/crashstats_base.html]: All of these changes were necessary so that you don't get this:
screen shot 2017-03-31 at 9 56 27 am

File: webapp-django/crashstats/crashstats/tests/test_views.py [hunk 2]
```diff
@@ -369,7 +369,7 @@ def mocked_product_versions(**params):
                     'throttle': '99.00',
                     'end_date': end_date,
                     'start_date': '2012-03-08',
-                    'is_featured': True,
+                    'is_featured': False,
```
Comment (peterbe) [inline webapp-django/crashstats/crashstats/tests/test_views.py]: I changed both of these ^ first to see if that broke any tests. It didn't. So I reused this piece of fixture for my new changes.

File: webapp-django/crashstats/crashstats/views.py [hunk 3]
```diff
@@ -510,9 +510,15 @@ def crashes_per_day(request, default_context=None):
...
         params['versions'] = []
-        for pv in context['active_versions'][context['product']]:
+        active_versions = context['active_versions'][context['product']]
+        for pv in active_versions:
             if pv['is_featured']:
                 params['versions'].append(pv['version'])
+        if not params['versions'] and active_versions:
+            # There were no featured versions, but there were active
+            # versions. Use the top 4 active versions instead.
+            for pv in active_versions[:4]:
```
Comment (peterbe) [inline webapp-django/crashstats/crashstats/views.py]: This number `4` *could* be moved to `settings` but I figured we'd never go into consul and change that. If we decide that 3 is the right number instead, we might as well just make another PR.

File: webapp-django/crashstats/topcrashers/views.py [hunk 4]
```diff
@@ -213,6 +213,15 @@ def topcrashers(request, days=None, possible_days=None, default_context=None):
...
                 return redirect(url)
+        if context['active_versions'][product]:
+            # Not a single version was featured, but there were active
+            # versions. In this case, use the first available
+            # *active* version.
+            for pv in context['active_versions'][product]:
+                url = '%s&version=%s' % (
+                    request.build_absolute_uri(), urlquote(pv['version'])
+                )
+                return redirect(url)
```
Comment (peterbe) [inline webapp-django/crashstats/topcrashers/views.py]: There is a bit of repeated code here but I think it's really really easy to read and there isn't much actual business logic that is repeated.

File: webapp-django/crashstats/crashstats/tests/test_views.py [hunk 5]
```diff
@@ -1102,6 +1102,59 @@ def is_percentage(thing):
...
+                    'release': 0.1,
+                    'beta': 1.0,
+                }
+            }
+
+        models.ProductBuildTypes.implementation().get.side_effect = (
+            mocked_product_build_types_get
+        )
+
+        models.ADI.implementation().get.side_effect = mocked_adi_get
```
Comment (adngdb) [inline webapp-django/crashstats/crashstats/tests/test_views.py]: Nit: this would be better right after the ``mocked_adi_get`` function imo.

File: webapp-django/crashstats/crashstats/views.py (hunk 3 above)
Comment (adngdb) [inline webapp-django/crashstats/crashstats/views.py]: It doesn't make sense to have that in ``settings``, but we have a weird way of handling constants (basically: we don't use them). This could very well go into a ``NUMBER_OF_FEATURED_VERSIONS`` constant at the top of this file.

File: webapp-django/crashstats/home/tests/test_views.py [hunk 6]
```diff
@@ -26,6 +26,20 @@ def test_home(self):
         ok_('WaterWolf 4.0.1' in response.content)
         ok_('WaterWolf 19.0' not in response.content)
 
+    def test_home_product_without_featued_versions(self):
```
Comment (adngdb) [inline webapp-django/crashstats/home/tests/test_views.py]: ``featued`` -> ``featured``

File: webapp-django/crashstats/home/views.py [hunk 7]
```diff
@@ -26,6 +26,15 @@ def home(request, product, default_context=None):
...
         ]
+        # If there are no featured versions but there are active
+        # versions, then fall back to use that instead.
+        if not context['versions'] and context['active_versions'][product]:
+            # But when we do that, we have to make a manual cut-off of
+            # the number of versions to return. So make it max 4.
+            context['versions'] = [
+                x['version']
+                for x in context['active_versions'][product]
+            ][:4]
```
Comment (adngdb) [inline webapp-django/crashstats/home/views.py]: This is another place the ``NUMBER_OF_FEATURED_VERSIONS`` constant could be of use.

**Comment by adngdb** [comment]
Also, another rebase is needed.

**Comment by peterbe** [comment]
I decided to make the number `4` a setting. It was just too nuts that it was written in the code twice in two different files.
